src/modules/rust.rs

```diff
@@ -258,14 +367,91 @@ fn format_rustc_version(rustc_version: &str, version_format: &str) -> Option<Str
     }
 }
 
+fn format_toolchain(toolchain: &str, default_host_triple: Option<&str>) -> String {
+    default_host_triple
+        .map(|triple| toolchain.trim_end_matches(&format!("-{}", triple)))
+        .unwrap_or(toolchain)
+        .to_owned()
+}
+
+fn format_rustc_version_verbose(stdout: &str, toolchain: Option<&str>) -> Option<(String, String)> {
+    let (mut release, mut host) = (None, None);
+    for line in stdout.lines() {
+        if line.starts_with("release: ") {
+            release = Some(line.trim_start_matches("release: "));
+        }
+        if line.starts_with("host: ") {
+            host = Some(line.trim_start_matches("host: "));
+        }
+    }
+    let (release, host) = (release?, host?);
+    let version = format_semver(release);
+    let toolchain = toolchain
+        .map(ToOwned::to_owned)
+        .unwrap_or_else(|| format!("{}", host));
+    Some((version, toolchain))
+}
+
+fn format_semver(semver: &str) -> String {
+    format!(
+        "v{}",
+        semver.find('-').map(|i| &semver[..i]).unwrap_or(semver)
+    )
+}
+
 #[derive(Debug, PartialEq)]
 enum RustupRunRustcVersionOutcome {
     RustcVersion(String),
-    ToolchainName(String),
+    ToolchainNotInstalled(String),
+    ToolchainUnknown,
     RustupNotWorking,
     Err,
 }
 
+#[derive(Default, Debug, PartialEq, Deserialize)]
+struct RustupSettings {
+    default_host_triple: Option<String>,
+    default_toolchain: Option<String>,
+    overrides: HashMap<PathBuf, String>,
+    version: Option<String>,
+}
+
+impl RustupSettings {
+    fn load(_context: &Context) -> Option<Self> {
+        let path = rustup_home().ok()?.join("settings.toml");
+        Self::from_toml_str(&fs::read_to_string(path).ok()?)
+    }
+
+    fn from_toml_str(toml_str: &str) -> Option<Self> {
+        let settings = toml::from_str::<Self>(toml_str).ok()?;
+        match settings.version.as_deref() {
+            Some("12") => Some(settings),
+            _ => {
+                log::warn!(
+                    r#"Rustup settings version is {:?}, expected "12""#,
+                    settings.version
+                );
+                None
+            }
+        }
+    }
+
+    fn default_host_triple(&self) -> Option<&str> {
+        self.default_host_triple.as_deref()
+    }
+
+    fn default_toolchain(&self) -> Option<&str> {
+        self.default_toolchain.as_deref()
+    }
+
+    fn lookup_override(&self, cwd: &Path) -> Option<String> {
+        self.overrides
+            .iter()
+            .find(|(dir, _)| cwd.starts_with(dir))
+            .map(|(_, name)| name.to_owned())
+    }
```

Comment:
This will need to integrate the changes from #3359 (longest match) and #3668 (proper prefix handling).